xocto/storage/storage.py

```diff
@@ -0,0 +1,1563 @@
+        bucket = self.buffers[self.bucket_name]
+        bucket[destination] = bucket[s3_object.key]
+        return S3Object(bucket_name=self.bucket_name, key=destination)
+
+    def rename(self, *, s3_object: S3Object, destination: str) -> S3Object:
+        assert s3_object.bucket_name == self.bucket_name
+        bucket = self.buffers[self.bucket_name]
+        if s3_object.key not in bucket:
+            raise KeyDoesNotExist("Key with path %s was not found" % s3_object.key)
+        if destination == s3_object.key:
+            return s3_object  # No rename required
+        bucket[destination] = bucket[s3_object.key]
+        del bucket[s3_object.key]
+        return S3Object(bucket_name=self.bucket_name, key=destination)
+
+    def delete(self, *, s3_object: S3Object) -> None:
+        bucket = self.buffers[self.bucket_name]
+        del bucket[s3_object.key]
+
+    def clear(self) -> None:
+        """
+        Deletes all files from all buckets, across all MemoryFileStore instances.
+        """
+        for bucket in self.buffers.values():
+            bucket.clear()
+
+    def generate_presigned_post(self, *, key_path: str, expires_in: int = 60) -> PreSignedPost:
+        return PreSignedPost(
+            # Resolves to a localdev/storage url
+            url=reverse("fake-presigned-post-upload"),
+            key=key_path,
+            # These fields should be passed in the body of the upload POST request by clients that wish to upload a file
+            # In production, they contain the signature used by AWS to verify the upload.
+            # Locally, however, we return the bucket name and key path so that when the upload request is made
+            # we know where to store the file.
+            fields={"key_path": key_path, "bucket_name": self.bucket_name},
+        )
+
+    def get_size_in_bytes(self, *, s3_object: S3Object) -> int:
+        assert s3_object.bucket_name == self.bucket_name
+        return len(self.fetch_file_contents(s3_object.key))
+
+
+def store(
+    bucket_name: str, use_date_in_key_path: bool = True, set_acl_bucket_owner: bool = False
+) -> BaseS3FileStore:
+    """
+    Return the appropriate storage instance for a given bucket.
+    """
+    storage_class = import_string(settings.STORAGE_BACKEND)
+    return storage_class(
+        bucket_name=bucket_name,
+        use_date_in_key_path=use_date_in_key_path,
+        set_acl_bucket_owner=set_acl_bucket_owner,
+    )
+
+
+def email_store(bucket_name: str) -> BaseS3FileStore:
+    """
+    Return the storage instance for email content.
+    """
+    store_class = import_string(settings.EMAIL_STORAGE_BACKEND)
+    return store_class(bucket_name)
+
+
+def fileserver_store(use_date_in_key_path: bool = True) -> BaseS3FileStore:
+    return store(settings.S3_FILESERVER_BUCKET, use_date_in_key_path)
+
+
+def flows_outbound_store() -> BaseS3FileStore:
+    return store(settings.S3_PRODUCTION_FLOWS_OUTBOUND)
```

Comment:
These factory functions are quite specific to KrakenCore - am not sure this should be in an open-source repo.